pkg/controller/persistentvolume/persistentvolume_framework_test.go

```diff
@@ -0,0 +1,462 @@
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	expectedMap := make(map[string]*api.PersistentVolumeClaim)
+	gotMap := make(map[string]*api.PersistentVolumeClaim)
+	for _, c := range expectedClaims {
+		c.ResourceVersion = ""
+		expectedMap[c.Name] = c
+	}
+	for _, c := range r.claims {
+		// We must clone the claim because of golang race check - it was
+		// written by the controller without any locks on it.
+		clone, _ := conversion.NewCloner().DeepCopy(c)
+		c = clone.(*api.PersistentVolumeClaim)
+		c.ResourceVersion = ""
+		gotMap[c.Name] = c
+	}
+	if !reflect.DeepEqual(expectedMap, gotMap) {
+		// Print ugly but useful diff of expected and received objects for
+		// easier debugging.
+		return fmt.Errorf("Claim check failed [A-expected, B-got result]: %s", diff.ObjectDiff(expectedMap, gotMap))
+	}
+	return nil
+}
+
+// popChange returns one recorded updated object, either *api.PersistentVolume
+// or *api.PersistentVolumeClaim. Returns nil when there are no changes.
+func (r *volumeReactor) popChange() interface{} {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if len(r.changedObjects) == 0 {
+		return nil
+	}
+
+	// For debugging purposes, print the queue
+	for _, obj := range r.changedObjects {
+		switch obj.(type) {
+		case *api.PersistentVolume:
+			vol, _ := obj.(*api.PersistentVolume)
+			glog.V(4).Infof("reactor queue: %s", vol.Name)
+		case *api.PersistentVolumeClaim:
+			claim, _ := obj.(*api.PersistentVolumeClaim)
+			glog.V(4).Infof("reactor queue: %s", claim.Name)
+		}
+	}
+
+	// Pop the first item from the queue and return it
+	obj := r.changedObjects[0]
+	r.changedObjects = r.changedObjects[1:]
+	return obj
+}
+
+// syncAll simulates the controller periodic sync of volumes and claim. It
+// simply adds all these objects to the internal queue of updates. This method
+// should be used when the test manually calls syncClaim/syncVolume. Test that
+// use real controller loop (ctrl.Run()) will get periodic sync automatically.
+func (r *volumeReactor) syncAll() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	for _, c := range r.claims {
+		r.changedObjects = append(r.changedObjects, c)
+	}
+	for _, v := range r.volumes {
+		r.changedObjects = append(r.changedObjects, v)
+	}
+	r.changedSinceLastSync = 0
+}
+
+func (r *volumeReactor) getChangeCount() int {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.changedSinceLastSync
+}
+
+func newVolumeReactor(client *fake.Clientset, ctrl *PersistentVolumeController, volumeSource, claimSource *framework.FakeControllerSource) *volumeReactor {
+	reactor := &volumeReactor{
+		volumes:      make(map[string]*api.PersistentVolume),
+		claims:       make(map[string]*api.PersistentVolumeClaim),
+		ctrl:         ctrl,
+		volumeSource: volumeSource,
+		claimSource:  claimSource,
+	}
+	client.AddReactor("*", "*", reactor.React)
+	return reactor
+}
+
+func newPersistentVolumeController(kubeClient clientset.Interface) *PersistentVolumeController {
```

Comment:
Good point. It started small and then it diverged as the unit test framework became more powerful. I'll make `NewPersistentVolumeController` more generic and rename `newPersistentVolumeController` to `newTestController` or so.